presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java

```diff
@@ -0,0 +1,240 @@
+    public int getPositionCount()
+    {
+        return valueIsNull.size();
+    }
+
+    @Override
+    protected Block getValues()
+    {
+        return values;
+    }
+
+    @Override
+    protected Slice getOffsets()
+    {
+        return offsets.getUnderlyingSlice();
+    }
+
+    @Override
+    protected Slice getValueIsNull()
+    {
+        return valueIsNull.getUnderlyingSlice();
+    }
+
+    @Override
+    public void assureLoaded()
+    {
+    }
+
+    private void checkOngoingBlockEntry()
+    {
+        if (ongoingBlockEntry) {
+            throw new IllegalStateException("Cannot write to BlockBuilder between calls to beginBlockEntry and closeEntry");
+        }
+    }
+
+    @Override
+    public BlockBuilder writeByte(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeShort(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeInt(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeLong(long value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeFloat(float value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeDouble(double value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeObject(Object value)
+    {
+        checkOngoingBlockEntry();
+
+        Block block = (Block) value;
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                values.appendNull();
+            }
+            else {
+                block.writePositionTo(i, values);
+                values.closeEntry();
+            }
+        }
+
+        if (currentEntrySize != 0) {
+            throw new IllegalStateException("Expected entry size to be exactly " + 0 + " but was " + currentEntrySize);
+        }
+        currentEntrySize++;
+        return this;
+    }
+
+    @Override
+    public ArrayElementBlockWriter beginBlockEntry()
+    {
+        checkOngoingBlockEntry();
+
+        if (currentEntrySize != 0) {
+            throw new IllegalStateException("Expected current entry size to be exactly 0 but was " + currentEntrySize);
+        }
+        currentEntrySize++;
+        return new ArrayElementBlockWriter(values, values.getPositionCount());
+    }
+
+    @Override
+    public BlockBuilder closeEntry()
+    {
+        if (currentEntrySize != 1) {
+            throw new IllegalStateException("Expected entry size to be exactly 1 but was " + currentEntrySize);
+        }
+
+        entryAdded(false);
+        ongoingBlockEntry = false;
```

Comment:
`ongoingBlockEntry` is never set to true, also I think you should remove `currentEntrySize` since it looks like it's superceded by `ongoingBlockEntry`